httpcore5/src/main/java/org/apache/hc/core5/http/io/ResponseOutOfOrderStrategy.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.io;
+
+import org.apache.hc.core5.http.ClassicHttpRequest;
+
+/**
+ * Represents a strategy to determine how frequently the client should check for an out of order response.
+ * An out of order response is sent before the server has read the full request. If the client fails to
+ * check for an early response then a {@link java.net.SocketException} or {@link java.net.SocketTimeoutException}
+ * may be thrown while writing the request entity after a timeout is reached on either the client or server.
+ *
+ * @since 5.1
+ */
+public interface ResponseOutOfOrderStrategy {
+
+    /**
+     * Returns the interval in bytes describing how often the socket is checked for an out-of-order response.
+     * A value of zero disables the check. This may result in requests blocking until the socket
+     * timeout is reached while a server sends an out-of-order response.
+     */
+    int checkInterval(ClassicHttpRequest request);
```

Comment:
Something else we may want to consider is the ability to check for an early response only once, after the first chunk or two has been sent. I think the most common case if for a server to parse the request and return immediately before consuming content when a precondition fails (a bad credential, for example).

This way we could provide most of the safety without sacrificing throughput for large uploads.

Reply:
This is a very good idea. This is what I have observed with Tomcat. Please also note that the first check might not succeed because BIO doesn't scale well. It make take upto 5 iterations to detect this. Sometimes even more. I don't know how to make this reliably. See my outout http://home.apache.org/~michaelo/issues/early-response/ accompanied with the known mailing list thread.